gtk: implement sensitive content reveal on paste confirmation

b6k-dev opened the pull request:
Fixes https://github.com/ghostty-org/ghostty/issues/4947 for gtk
This PR implements the senstive content hiding when displaying the paste confirmation dialog in secure input mode.

Following changes are implemented:
- in the blueprint for each dialog add a show/hide button that is not visible by default, and a Revealer that is revealed by default
- save the `secure_input` action value for each surface in the GTK apprt
- pass the value when initializing the paste confirmation dialog
- in the dialog code, alter the visibility of the content and reveal/hide buttons based on secure input flag value

Demo:

https://github.com/user-attachments/assets/c91cbd3d-ed3b-464d-b4cf-e51fe7aa23b7

I feel like this is already a nearly full implementation, but I'm leaving this as a draft for now, since i need to look into blueprints for Adwaita 1.2, and verify if it behaves properly when the dialog is in not-sensitive input mode and in OSC52 mode.

b6k-dev committed:
gtk: implement sensitive content reveal mechanism when showing paste confirmation in secure input mode

jcollie commented on src/apprt/gtk/ui/1.5/ccw-osc-52-read.blp:
```diff
@@ -14,18 +14,46 @@ Adw.AlertDialog clipboard_confirmation_window {
+      label: _("Hide Potentially Sensitive Content");
+    }
+
+    Revealer content_revealer {
```
I would prefer an overlay that blocks/blurs the content rather than a revealer. I dislike how the window size jumps around when you reveal/hide the content.

maciekbartczak replied:
Switched to the blur approach, I agree it looks much better when the window does not jump around.

https://github.com/user-attachments/assets/40d6d1cd-bcd4-4892-aa1e-030b28226ae4

jcollie commented on src/apprt/gtk/Surface.zig:
```diff
@@ -1203,12 +1207,7 @@ fn gtkClipboardRead(
-                &self.core_surface,
-                req.state,
-            ) catch |window_err| {
+            ClipboardConfirmationWindow.create(self.app, str, &self.core_surface, req.state, self.is_secure_input) catch |window_err| {
```
Make sure to keep the trailing `,` so that `zig fmt` wraps this line.

jcollie commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2205,12 +2204,7 @@ fn doPaste(self: *Surface, data: [:0]const u8) void {
-                &self.core_surface,
-                .paste,
-            ) catch |window_err| {
+            ClipboardConfirmationWindow.create(self.app, data, &self.core_surface, .paste, self.is_secure_input) catch |window_err| {
```
Make sure to keep the trailing `,` so that `zig fmt` wraps this line.

jcollie commented on src/apprt/gtk/ClipboardConfirmationWindow.zig:
```diff
@@ -25,12 +25,16 @@ dialog: *DialogType,
 pending_req: apprt.ClipboardRequest,
+content_revealer: *gtk.Revealer,
+reveal_button_widget: *gtk.Widget,
+hide_button_widget: *gtk.Widget,
```
The buttons should saved as `*gtk.Button` and cast as needed later.

jcollie reviewed (changes requested):
As noted in the start of the PR, need to get this working for Adwaita 1.2.

b6k-dev committed:
gtk: blur the content view instead of using a Revealer widget

pluiedev commented on src/apprt/gtk/ClipboardConfirmationWindow.zig:
```diff
@@ -88,15 +99,39 @@ fn init(
     text_view.setBuffer(buffer);
 
+    if (is_secure_input) {
+        gobject.ext.as(gtk.Widget, text_view_scroll).setSensitive(@intFromBool(false));
+        gobject.ext.as(gtk.Widget, self.text_view).addCssClass("blurred");
```
You can write as in postfix position: `foo.as(gtk.Bar)`

pluiedev reviewed (changes requested):
From a UI design perspective I'm not sure if a background blur is idiomatic for GTK4 apps, although I'm not opposed to it. A separate row button is definitely not it, though... maybe an overlaid button in the corner could be used to reenable the overlay? I'm not sure why anyone would do this anyway.

tristan957 commented:
The blur is sweet

jcollie commented:
I like the blur as well, but I agree that a separate button isn't the right UI. Generally clicking on the blurred panel is what I see for UI.

jcollie commented:
> How about something like this?

Looks good!

> I'm not really sure what to think about the hide button, it feels out of place in there, maybe it is actually not needed at all?

Yeah, I'd just get rid of the hide button, and let clicking on the text view itself hide the content again.

tristan957 commented:
What about a button similar to a password entry. Just and :eye: button overlaid in the top right corner (or top left corner) in LTR languages?

pluiedev commented:
> Just an :eye: button overlaid in the top right corner 

This

b6k-dev committed:
gtk: new approach to reveal/hide buttons

maciekbartczak commented:
Ok, that looks perfect to me, but I'm open to any suggestions :smile: 

https://github.com/user-attachments/assets/3c63765d-edad-4d94-900f-295e730d9282

jcollie commented:
> Ok, that looks perfect to me, but I'm open to any suggestions 😄

Looks good!

tristan957 commented on src/apprt/gtk/ui/1.5/ccw-paste.blp:
```diff
@@ -10,22 +10,62 @@ Adw.AlertDialog clipboard_confirmation_window {
+      margin-end: 12;
+      margin-top: 12;
+
+      styles [
+        "destructive-action",
+      ]
```
I would not use suggested or destructive action styles for the overlay. Just the icon in regular colors will be fine. The dialog already has suggested and destructive coloring. Seems a little overwhelming

b6k-dev committed:
code review:
- implement blueprints for Adwaita 1.2
- use postifx notation for casting gtk widgets
- fix formatting

maciekbartczak commented:
Ok, that should be the final version. I've Added the implemenation for Adwaita 1.2, addressed the other issues and resolved merge conflicts.

jcollie commented:
Thanks @maciekbartczak for putting up with all our nitpicks, great work!
